synapse/handlers/message.py

```diff
@@ -1001,13 +1001,52 @@ async def create_new_client_event(
             )
 
         self.validator.validate_new(event, self.config)
+        await self._validate_event_relation(event)
+        logger.debug("Created event %s", event.event_id)
+
+        return event, context
+
+    async def _validate_event_relation(self, event: EventBase) -> None:
+        """
+        Ensure the relation data on a new event is not bogus.
+
+        Args:
+            event: The event being created.
+
+        Raises:
+            SynapseError if the event is invalid.
+        """
+
+        relation = event.content.get("m.relates_to")
+        if not relation:
+            return
+
+        relation_type = relation.get("rel_type")
+        if not relation_type:
+            return
+
+        # Ensure the parent is real.
+        relates_to = relation.get("event_id")
+        if not relates_to:
+            return
+
+        parent_event = await self.store.get_event(relates_to, allow_none=True)
+        if parent_event:
+            # And in the same room.
+            if parent_event.room_id != event.room_id:
+                raise SynapseError(400, "Relations must be in the same room")
+
+        else:
+            # There must be some reason that the client knows the event exists,
+            # see if there are existing relations. If so, assume everything is fine.
+            if not await self.store.event_related_to(relates_to):
+                # Otherwise, the client can't know about the parent event!
+                raise SynapseError(400, "Can't send relation to unknown event")
```

Comment:
> Is it because you might see a thread where the base message is not visible to you/your homeserver, but the subsequent messages are?

This is the situation that I was thinking of, yes. (The simple case is if you're invited to a room that doesn't share history and you can't see the thread head, but you can see subsequent messages in it).

I think it is OK to divulge in this case that the event ID is valid. Theoretically you would know already since you've seen other events referencing that event.

It could be used to iterate through whether events are in a room, but that seems unreasonable in terms of the space to scan + rate limits, I think.

Reply:
Yeah, that seems OK. I don't think we're particularly stringent about this anyway. I can't properly imagine a case where this would hurt us, but that doesn't mean it's not possible (though we probably already divulge this elsewhere)